src/libertem/udf/masks.py

```diff
@@ -64,34 +47,65 @@ def process_flat(flat_tile):
             and 'sparse.pydata' in self.masks.use_sparse
         ):
+            self.process_flat = self._process_flat_sparsepyd
+            self.needs_transpose = False
+        else:
+            self.process_flat = self._process_flat_standard
 
-            def process_flat(flat_tile):
-                masks = self.masks.get_for_sig_slice(
-                    self.meta.sig_slice, transpose=False
-                )
-                # Make sure the sparse.pydata mask comes first
-                # to choose the right multiplication method
-                result = (masks @ flat_tile.T).T
-                return result
+    def _get_masks(self):
+        return self.masks.get_for_sig_slice(
+            self.meta.sig_slice, transpose=self.needs_transpose
+        )
 
-        else:
+    def _process_flat_torch(self, flat_tile, masks):
+        import torch
+        # CuPy back-end disables torch in get_task_data
+        # FIXME use GPU torch with CuPy array?
+        return torch.mm(
+            torch.from_numpy(flat_tile),
+            torch.from_numpy(masks),
+        ).numpy()
 
-            def process_flat(flat_tile):
-                masks = self.masks.get_for_sig_slice(
-                    self.meta.sig_slice, transpose=True
-                )
+    def _process_flat_spsp(self, flat_tile, masks):
+        return rmatmul(flat_tile, masks)
 
-                result = flat_tile @ masks
-                return result
+    def _process_flat_sparsepyd(self, flat_tile, masks):
+        # Make sure the sparse.pydata mask comes first
+        # to choose the right multiplication method
+        return (masks @ flat_tile.T).T
 
-        self.process_flat = process_flat
+    def _process_flat_standard(self, flat_tile, masks):
+        return flat_tile @ masks
 
     def process_tile(self, tile):
         flat_shape = (tile.shape[0], prod(tile.shape[1:]))
         # Avoid reshape since older versions of scipy.sparse don't support it
         flat_data = tile.reshape(flat_shape) if tile.shape != flat_shape else tile
+        return self.process_flat(flat_data, self._get_masks())
+
+    def process_frame_shifted(self, frame, shifts: Tuple[int, ...]):
+        masks = self._get_masks()
+        num_masks = len(self.masks)
+        shifted_slice = self.meta.sig_slice.shift_by(shifts)
+        left, right = self.meta.sig_slice.intersection_pair(shifted_slice)
+        if left.is_null():
+            # Zero overlap after shifts, shortcut return
+            return np.full((num_masks,), np.nan, dtype=np.float32)
+        mask_slice = right.get()
+        if self.needs_transpose:
+            mask_slice = mask_slice + (slice(None), )
+            masks = masks.reshape((*frame.shape, num_masks))
+            final_mask_shape = (-1, num_masks)
+        else:
+            mask_slice = (slice(None), ) + mask_slice
+            masks = masks.reshape((num_masks, *frame.shape))
+            final_mask_shape = (num_masks, -1)
```

Comment:
I think this is missed because I never test with a specific scipy.sparse array backend + shifts.

Your observation that 'an "identity"/zero shifts' vector should match the standard route is a good one, I will try this.

Reply:
This was a bit of a rabbit hole! I finally managed to pass `scipy.sparse` frames into`ApplyMasksEngine.process_frame_shifted` (had to set `backends=(backend,)` in the args of `run_udf`, which results in `needs_transpose == False` (as long as `use_sparse='sparse.pydata'` at the same time (😅!!).

We can get the right result here, but the shifting logic needs a frame with `sig_shape`, rather than the default `(1, sig_size)` shape that `array_backend=BACKEND_SCIPY_CSR` normally gives us. This can be done with a reshape (assuming scipy sufficiently recent), but at least for CSR this casts the array to COO, which then can't be sliced... So if I cast again back to CSR, the rest of the function runs correctly and passes the tests (modulo the bugs in the non-transposed reshaping that I've fixed).

So, the question I have ... just because it's technically possible, do we want to support all these gynmastics to support `array_backend=scipy.sparse, use_sparse=sparse.pydata, shifts is not None` ?

One alternative would be to only support `sparse.pydata` array backends if using `shifts`. This substantially reduces the reshape/casting overhead. What I don't know is if supporting scipy sparse backends are vital for Timepix data, for example...

Reply:
In general, I would think supporting shifts with usable performance on sparse input data is worthwhile, but figuring out which method might need a small benchmark.

> One alternative would be to only support `sparse.pydata` sparse backends if using `shifts`

I currently don't have an intuition which of the options would perform better. `sparse.pydata` is rather slow for `dot` IIRC, compared to `CSR @ CSC`, but if there are multiple conversion steps over the whole data, it might be faster than doing the gymnastics (I think most conversions and slice operations allocate new arrays behind the scenes! Except for some same-type different-library conversions. Need to keep that in mind...).

> What I don't know is if supporting scipy sparse backends are vital for Timepix data, for example...

If your UDF signals that it only supports `sparse.pydata` as input, an execution plan will be built up that runs `sparseconverter` on the input data before passing it to your UDF. The question is really, how do the conversion steps of the gymnastics compare to the lower performance of COO `dot`?

Reply:
So I have implemented scipy CSR and CSC array backend support for shifts, and scipy.sparse masks are required to be converted to sparse.pydata equivalents. The prospect of setting up benchmarks for all of the combinations is a bit daunting right now, I think I need to come back to that another day with a fresh look!

Reply:
The other component of this is whether CuPy worker + sparse mask + `use_sparse=None` => densification of the mask should trigger densification of the frame as part of the execution plan, too, i.e. we drop sparse array backends conditional on all of the above. Firstly I can't figure out a way to implement this, because I think the execution plan is built once on the main node and the UDF doesn't get to help construct its plan based on different worker types (this is the subject of the conversation we had the other week). Secondly I actually don't know if it's right do do this, because I would guess in certain cases, indeed often, that `(sparse_frame @ dense_mask)` could run faster than `(dense_frame @ dense_mask)`. As far as I can tell the test cases cover all the combinations we could expect to receive, but without doing some monkeypatching to register which combinations are actually received in `ApplyMasksEngine` I am not 100% sure yet.

Reply:
These are the current combinations that arrive into the `process_shifted_frame` function in `ApplyMasksUDF` with the main shifted test:

```shell
Frame                                    Mask
cupy.ndarray                             ['cupy.ndarray']
cupyx.scipy.sparse._csc.csc_matrix       ['cupy.ndarray']
cupyx.scipy.sparse._csr.csr_matrix       ['cupy.ndarray']
numpy.ndarray                            ['sparse._coo.core.COO', 'numpy.ndarray']
scipy.sparse._csr.csr_matrix             ['sparse._coo.core.COO', 'numpy.ndarray']
sparse._compressed.compressed.GCXS       ['sparse._coo.core.COO', 'numpy.ndarray']
sparse._coo.core.COO                     ['sparse._coo.core.COO', 'numpy.ndarray']
```

So at least we're certain the sparse frame / dense mask combination works on CuPy